File: LEETCODE/JAVA/ThreeSum.java

import java.util.*;

public class ThreeSum{
    public static List<List<Integer>> ThreeSumFunc(int[] nums){
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(nums);
        for(int i = 0; i < nums.length; i++){
            if(i > 0 && nums[i] == nums[i-1]) continue;
            int j = i + 1;
            int k = nums.length - 1;
            while(j < k){
                int total = nums[i] + nums[j] + nums[k];
                if(total > 0){
                    j++;
                }else if(total < 0){
                    k--;
                }else{
                    result.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;
                    while(j < k && nums[j] == nums[j-1]) j++;
                    while(j < k && nums[k] == nums[k+1]) k--;
                }
            }
        }
        return result;
    }
    public static void main(String[] args) {
        int[] nums = {-1, 0, 1, 2, -1, -4};
        List<List<Integer>> result = ThreeSumFunc(nums);
        System.out.println(result);
    }
    
}

File: LEETCODE/PYTHON/ThreeSum.py
# 15. 3Sum Level - (Medium) Topic -(Array ,Two Pointers,Sorting)

def threeSumFunc(nums):
    result = []
    nums.sort()

    for i in range(len(nums)):
        if(i > 0 and nums[i] == nums[i-1]): continue
        j = i + 1
        k = len(nums) - 1
        while(j < k):
            total = nums[i] + nums[j] + nums[k];
            if total < 0:
                j += 1
            elif total > 0;
                k -= 1
            else:
                result.append([nums[i], nums[j], nums[k]])
                k -= 1
                j += 1
                while(j < k and nums[j] == nums[j-1]): j += 1
                while(j < k and nums[k] == nums[k+1]): k += 1
        return result


nums = [-1, 0, 1, 2, -1, -4]
print(threeSumFunc(nums))
